db_sql.go

```diff
@@ -469,89 +580,161 @@ func SqlCreateDatabase(conx *DBConnection) error {
+	// and for running on brand new systems.
+	upToDate, err := runBootstrap(conx)
+	if err != nil {
+		return false, err
+	}
 
-func migrateSchemaInternal(tryCreatingDB bool, conx *DBConnection, schema_name string, migrations []string) (migrateError error) {
-	if db, eConnect := conx.Connect(); eConnect != nil {
-		return eConnect
-	} else {
-		defer db.Close()
+	if upToDate == true {
+		conStr := fmt.Sprintf("host=%v user=%v password=%v dbname=%v sslmode=disable", conx.Host, conx.User, conx.Password, conx.Database)
+		_, err := migration.Open(conx.Driver, conStr, createMigrations())
+		if err != nil {
+			return false, err
+		}
+		return true, err
+	}
 
-		if tx, eTxBegin := db.Begin(); eTxBegin == nil {
-			defer func() {
-				if err := recover(); err == nil {
-					migrateError = tx.Commit()
-				} else {
-					tx.Rollback()
-					migrateError = err.(error)
-				}
-			}()
+	return false, nil
+}
 
-			version, eGetVersion := readSchemaVersion(tx, schema_name)
-			if eGetVersion != nil {
-				panic(eGetVersion)
-			}
+func createVersionTable(db *sql.DB, version int) error {
+	_, err := db.Exec(fmt.Sprintf(`
+		CREATE TABLE migration_version (
+			version INTEGER
+		);
+		INSERT INTO migration_version (version) VALUES (%v)`, version))
+	return err
+}
 
-			if version > len(migrations) {
-				panic(errors.New(fmt.Sprintf("%v database is newer than this program (%v)", schema_name, version)))
-			}
+func createMigrations() []migration.Migrator {
+	var migrations []migration.Migrator
+
+	text := []string{
+		// authgroup migrations
+		`CREATE TABLE authgroup (id SERIAL PRIMARY KEY, name VARCHAR, permlist VARCHAR);
+		CREATE UNIQUE INDEX idx_authgroup_name ON authgroup (name);`,
+
+		// authsession migrations
+		`CREATE TABLE authsession (id BIGSERIAL PRIMARY KEY, sessionkey VARCHAR, identity VARCHAR, permit VARCHAR, expires TIMESTAMP);
+		CREATE UNIQUE INDEX idx_authsession_token ON authsession (sessionkey);
+		CREATE INDEX idx_authsession_identity ON authsession (identity);
+		CREATE INDEX idx_authsession_expires  ON authsession (expires);`,
+
+		`DELETE FROM authsession;`,
+
+		// authuser migrations
+		`CREATE TABLE authuser (id BIGSERIAL PRIMARY KEY, identity VARCHAR, password VARCHAR, permit VARCHAR);
+		CREATE UNIQUE INDEX idx_authuser_identity ON authuser (identity);`,
+
+		`DROP INDEX idx_authuser_identity;
+		CREATE UNIQUE INDEX idx_authuser_identity ON authuser (LOWER(identity));`,
+
+		`ALTER TABLE authuser ADD COLUMN pwdtoken VARCHAR;`,
+
+		// Change from using email address as the primary identity of a user, to a 64-bit integer, which we call UserId.
+		// The above migrations were from the old migration system, from this migration on we are using the new migration
+		// system from BurntSushi
+		`CREATE TABLE authuserstore (userid BIGSERIAL PRIMARY KEY, email VARCHAR, username VARCHAR, firstname VARCHAR, lastname VARCHAR, mobile VARCHAR, archived BOOLEAN default false);
```

Comment:
I don't think we should rely on DEFAULT db values (ie used here for the archived field). We should rather be explicit. It's like static type checking. If we don't have any DEFAULT value, then we will see a NULL value for archived, and know that we've forgotten to set it somewhere.